preprocessing/structs.ts

```diff
@@ -247,39 +251,43 @@ library OfferDetailUnpackedExtra {
+      /* * `density` is similar to a 'dust' parameter. We prevent spamming of low-volume offers by asking for a minimum 'density' in `outbound_tkn` per gas requested. For instance, if `density` is worth 10, `offer_gasbase == 5000`, an offer with `gasreq == 30000` must promise at least _10 × (30000 + 5000) = 350000_ `outbound_tkn`. _9 bits wide_.
 
-      We store the density as a float with 2 bits for the mantissa, 7 for the exponent, and an exponent bias of 32, so density ranges from $2^{-32}$ to $1.75 \times 2^{95}$. For more information, see `DensityLib`.
+      We store the density as a float with 2 bits for the mantissa, 7 for the exponent, and an exponent bias of 32, so that density ranges from $2^{-32}$ to $1.75 \times 2^{95}$. For more information, see `DensityLib`.
       
       */
       { name: "density", bits: 9, type: "Density", underlyingType: "uint"},
+      /* To save on gas, Mangrove cache the entire tick tree branch of the bin that contains the best offer in each offerList's `local` parameter. Taken together, `binPosInLeaf`, `level3`, `level2`, `level1` and `root` provide the following info:
+      - What the current bin is (see `BinLib.bestBinFromLocal`)
+      - When a leaf is emptied and the next offer must be fetched, the information in the fields `level3`, `level2`, `level1` and `root` avoid multiple storage reads
+      */
       { name: "binPosInLeaf", bits: 2, type: "uint" },
       { name: "level3", bits: 64, type: "Field", underlyingType: "uint" },
       { name: "level2", bits: 64, type: "Field", underlyingType: "uint" },
       { name: "level1", bits: 64, type: "Field", underlyingType: "uint" },
       { name: "root", bits: 2, type: "Field", underlyingType: "uint" },
-      /* * `offer_gasbase` is an overapproximation of the gas overhead associated with processing one offer. The Mangrove considers that a failed offer has used at least `offer_gasbase` gas. The actual field name is `kilo_offer_gasbase` and the accessor `offer_gasbase` returns `kilo_offer_gasbase*1e3`. Local to a pair, because the costs of calling `outbound_tkn` and `inbound_tkn`'s `transferFrom` are part of `offer_gasbase`. Should only be updated when ERC20 contracts change or when opcode prices change. */
+      /* * `offer_gasbase` is an overapproximation of the gas overhead associated with processing one offer. The Mangrove considers that a failed offer has used at least `offer_gasbase` gas. The actual field name is `kilo_offer_gasbase` and the accessor `offer_gasbase` returns `kilo_offer_gasbase*1e3`. Local to an offerList, because the costs of calling `outbound_tkn` and `inbound_tkn`'s `transferFrom` are part of `offer_gasbase`. Should only be updated when ERC20 contracts change or when opcode prices change. */
```

Comment:
Could we be a bit more specific - processing one offer sounds like if a market order has 10 offers this should count what one can consume. However, to compensate a taker of a single large failing offer it needs to compensate for the entire market order TX - or to incentivize cleaners, to compensate for at least a complete clean TX.

Reply:
Actually, the description in offer_Detail is better:
```
offer_gasbase represents the gas overhead used by processing the offer inside Mangrove + the overhead of initiating an entire order, in 1k gas increments.
```